crates/uv-publish/src/lib.rs

```diff
@@ -0,0 +1,784 @@
+use base64::prelude::BASE64_STANDARD;
+use base64::Engine;
+use distribution_filename::{DistFilename, SourceDistExtension, SourceDistFilename};
+use fs_err::File;
+use futures::TryStreamExt;
+use glob::{glob, GlobError, PatternError};
+use itertools::Itertools;
+use python_pkginfo::Metadata;
+use reqwest::header::AUTHORIZATION;
+use reqwest::multipart::Part;
+use reqwest::{Body, Response, StatusCode};
+use reqwest_middleware::RequestBuilder;
+use serde::Deserialize;
+use sha2::{Digest, Sha256};
+use std::collections::HashSet;
+use std::io::BufReader;
+use std::path::{Path, PathBuf};
+use std::{fmt, io};
+use thiserror::Error;
+use tokio::io::AsyncReadExt;
+use tracing::{debug, enabled, trace, Level};
+use url::Url;
+use uv_client::BaseClient;
+use uv_fs::Simplified;
+use uv_metadata::read_metadata_async_seek;
+
+#[derive(Error, Debug)]
+pub enum PublishError {
+    #[error("Invalid publish path: `{0}`")]
+    Pattern(String, #[source] PatternError),
+    /// [`GlobError`] is a wrapped io error.
+    #[error(transparent)]
+    Glob(#[from] GlobError),
+    #[error("Path patterns didn't match any wheels or source distributions")]
+    NoFiles,
+    #[error(transparent)]
+    Fmt(#[from] fmt::Error),
+    #[error("File is neither a wheel nor a source distribution: `{}`", _0.user_display())]
+    InvalidFilename(PathBuf),
+    #[error("Failed to publish: `{}`", _0.user_display())]
+    PublishPrepare(PathBuf, #[source] PublishPrepareError),
+    #[error("Failed to publish `{}` to `{}`", _0.user_display(), _1)]
+    PublishSend(PathBuf, Url, #[source] PublishSendError),
+}
+
+/// Failure to get the metadata for a specific file.
+#[derive(Error, Debug)]
+pub enum PublishPrepareError {
+    #[error(transparent)]
+    PkgInfoError(#[from] python_pkginfo::Error),
+    #[error(transparent)]
+    Io(#[from] io::Error),
+    #[error("Failed to read metadata")]
+    Metadata(#[from] uv_metadata::Error),
+    #[error("Only files ending in `.tar.gz` are valid source distributions: `{0}`")]
+    InvalidExtension(SourceDistFilename),
+    #[error("No PKG-INFO file found")]
+    MissingPkgInfo,
+    #[error("Multiple PKG-INFO files found: `{0}`")]
+    MultiplePkgInfo(String),
+    #[error("Failed to read: `{0}`")]
+    Read(String, #[source] io::Error),
+}
+
+/// Failure in or after (HTTP) transport for a specific file.
+#[derive(Error, Debug)]
+pub enum PublishSendError {
+    #[error("Failed to send POST request")]
+    ReqwestMiddleware(#[from] reqwest_middleware::Error),
+    #[error("Upload failed with status {0}")]
+    StatusNoBody(StatusCode, #[source] reqwest::Error),
+    #[error("Upload failed with status code {0}: {1}")]
+    Status(StatusCode, String),
+    /// The registry returned a "403 Forbidden"
+    #[error("Incorrect credentials (status code {0}): {1}")]
+    IncorrectCredentials(StatusCode, String),
```

Comment:
Per my earlier comment, maybe this should be "PermissionDenied" instead of "IncorrectCredentials"?